### PR TITLE
Migration to Gradle 8 and AGP 8.1.1

### LINKED ISSUE #89: Gradle 8 support
The builds are now using Gradle 7.x, which is current as of AS Giraffe 2022.3.1+. 

There is Gradle 8 already on the horizon and it's known to cause a variety of issues when upgrading regarding required changes to build flows.

This should be upgraded as soon as possible.

### PR DESCRIPTION
**Pull request by TheRealAgentK**
### Description :memo:

Fixes #89 and migrated the build over to Gradle 8 and AGP 8.1.1

Also fixes a bug where the sample app was outputting the incorrect provider version

**Type of change**

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

**Updates**

- Build to Gradle 8+
- AGP to 8.1.1
- Java target to Java 17 as required by Gradle 8
- UI bug fix in sample app re provider version

### Test plan :test_tube:

Project shouldn't trigger "Upgrade Gradle to 8" any more. Builds should all still work across `debug` and `release` flavours.

Note: Deprecation warning for Gradle 9+ still trigger and will be resolved separately.

### Author to check :eyeglasses:
- [x] Project and all contained modules build
- [x] Tested on appropriate devices/emulators and SDK versions

~- [] Reviewed by another developer~
N/A
- [x] Appropriate documentation written (code comments, internal docs)

### Reviewer to check :heavy_check_mark:
- [ ] Change has been tested on a device/emulator
- [ ] Code is written to standards

~- [ ] Appropriate tests have been written (code comments, internal docs)~
N/A

### CODE REVIEW

File: provider/build.gradle [hunk 1]
```diff
@@ -6,22 +6,20 @@ android {
     defaultConfig {
         minSdkVersion 16
         targetSdkVersion 34
-        versionCode = Long.valueOf(VERSION_CODE)
-        versionName = VERSION_NAME
```
Comment (sumitramanga) [inline provider/build.gradle:10]: I'm guessing we don't need to define his within the new version of Gradle?

File: app/build.gradle [hunk 2]
```diff
@@ -105,8 +105,8 @@ android {
         viewBinding true
     }
     compileOptions {
-        sourceCompatibility JavaVersion.VERSION_1_8
-        targetCompatibility JavaVersion.VERSION_1_8
+        sourceCompatibility JavaVersion.VERSION_17
```
Comment (sumitramanga) [inline app/build.gradle:114]: Is this change from 1.8 to 17?!

File: gradle.properties [hunk 3]
```diff
@@ -40,7 +40,7 @@ VERSION_NAME=4.1-SNAPSHOT
 # 4.0.0-alpha1 -> 40000031
 # 5.2.3        -> 50203000
 # 5.3.3-beta2  -> 50303072
-VERSION_CODE=40100031
+VERSION_CODE=40100032
```
Comment (sumitramanga) [inline gradle.properties:45]: Can we add a comment above this like how we have previously please?

File: provider/build.gradle (hunk 1 above)
Comment (TheRealAgentK) [reply provider/build.gradle:10]: It's just moved for clarity and consistency.

When you look further down in `buildTypes`, the `buildConfigField` lines don't refer to `defaultConfig...` anymore but reference the variables directly.

File: app/build.gradle (hunk 2 above)
Comment (TheRealAgentK) [reply app/build.gradle:114]: Lol, yup. BUT: Java's naming system changed in between Java 1.8 is essentially Java 8 in the current naming system, with 8, 11 and 17 LTS releases. Not sure what happened to 14 and why that never became an LTS.

File: gradle.properties (hunk 3 above)
Comment (TheRealAgentK) [reply gradle.properties:45]: Ah, sure. the comments above were always just examples re how to build the VERSION_CODE from a more readable VERSION_NAME. But we might as well use what we do as examples.

**Comment by sumitramanga** [comment]
Yay! image

**Comment by sumitramanga** [comment]
Will you be upgrading to the latest AGP in a later PR?

**Comment by sumitramanga** [comment]
Optional: It would be nice to have some feedback upon clicking the buttons within the providers test app. Such as, if you click there button, it should output some text that says 'Message sent' or 'Error sending message, check API key is valid'

**Comment by TheRealAgentK** [comment]
> Will you be upgrading to the latest AGP in a later PR?

Yeah, the AGP versions are somewhat tied to AS versions. 

I think we should update when new AS versions are released. Hedgehog is currently in RC and Iguana in Canary - when Hedgehog comes out of RC, we'll bump AGP up accordingly.

**Comment by TheRealAgentK** [comment]
> Optional: It would be nice to have some feedback upon clicking the buttons within the providers test app. Such as, if you click there button, it should output some text that says 'Message sent' or 'Error sending message, check API key is valid'

Great suggestion:

https://github.com/MindscapeHQ/raygun4android/issues/98

**Comment by sumitramanga** [comment]
> > Will you be upgrading to the latest AGP in a later PR?
> 
> Yeah, the AGP versions are somewhat tied to AS versions.
> 
> I think we should update when new AS versions are released. Hedgehog is currently in RC and Iguana in Canary - when Hedgehog comes out of RC, we'll bump AGP up accordingly.

I had to do a bit of reading to understand this 😄 It all makes sense now. Thank you! Android has so many concepts to keep up with. Does this change have an impact on the customer side or does it affect just developers of the provider

**Comment by sumitramanga** [comment]
Testing of the app went well. It sends CR and RUM data after rebuilding the project 👏

**Comment by sumitramanga** [comment]
One more comment before I approve. Some of these changes have been added automatically due to the upgrade, right? I had attempted to upgrade the AGP just to see what happens and that's what happened :)

**Review by sumitramanga — APPROVED** [review]
LGTM! 🤘

**Comment by TheRealAgentK** [comment]
> One more comment before I approve. Some of these changes have been added automatically due to the upgrade, right? I had attempted to upgrade the AGP just to see what happens and that's what happened :)

Yeah, the wizard from 7.4to 8.1 gets you some way there luckily, but it's patchy and inconsistent as well.

**Comment by TheRealAgentK** [comment]
> LGTM! 🤘

I'll make a few tweaks tomorrow re the small things we discussed and then merge it.

**Pull request merged**